### ISSUE TITLE
Setting of Parameters in gym_anytrading and How to Evaluate Training Performance

### ISSUE THREAD

**Issue by 123xian123** [opendilab/DI-engine#694]
Thank you very much for sharing the code. I used my own data for training, which consists of futures' intraday data, with "price" representing the settlement price for each minute. I utilized a large amount of data (I'm not sure if more data is better). I noticed there are many parameters in "stocks_dqn_config," but I'm not familiar with the meaning of each parameter. I have made several attempts to configure the parameters, as shown in the code below, and I hope to receive some advice on setting these parameters (the current settings require a lot of memory and often lead to memory overflow).

Additionally, I'm not sure how to determine whether the model's training has produced good results. Below are some initial training result charts (it seems that "eps_length" might be set too large), where the profit remains constantly at 1, and all the points in the price chart are "short" points (perhaps due to the early stage of training). I would greatly appreciate any help or guidance you can provide regarding the aforementioned questions.
`from easydict import EasyDict

stocks_dqn_config = dict(
    exp_name='stocks_dqn_seed0',
    env=dict(
        # Whether to use shared memory. Only effective if "env_manager_type" is 'subprocess'
        # Env number respectively for collector and evaluator.
        collector_env_num=16,
        evaluator_env_num=16,
        env_id='methanol_futures-v0',
        n_evaluator_episode=16,
        stop_value=2,
        # one trading year.
        eps_length=20700,
        # associated with the feature length.
        window_size=100,
        # the path to save result image.
        save_path='./fig/',
        # the raw data file name
        methanol_futures_data_filename='METHANOL_FUTURES',
        # the stocks range percentage used by train/test.
        # if one of them is None, train & test set will use all data by default.
        train_range=None,
        test_range=None,
    ),
    policy=dict(
        # Whether to use cuda for network.
        cuda=True,
        model=dict(
            # obs_shape=1602,
            # action_shape=3,
            # encoder_hidden_size_list=[32,128],
            # head_layer_num=2,
            obs_shape=802,
            action_shape=3,
            encoder_hidden_size_list=[256],
            head_layer_num=1,
            # Whether to use dueling head.
            dueling=True,
        ),
        # Reward's future discount factor, aka. gamma.
        discount_factor=0.99,
        # How many steps in td error.
        nstep=5,
        # learn_mode config
        learn=dict(
            update_per_collect=100,
            batch_size=128,
            learning_rate=0.001,
            # Frequency of target network update.
            target_update_freq=1000,
            ignore_done=True,
            # learner = dict(
            #     hook=dict(
            #         load_ckpt_before_run='.stocks_dqn_seed0/ckpt/ckpt_best.pth.tar',))
        ),
        # collect_mode config
        collect=dict(
            # You can use either "n_sample" or "n_episode" in collector.collect.
            # Get "n_sample" samples per collect.
            n_sample=3200,
            # Cut trajectories into pieces with length "unroll_len".
            unroll_len=1,
        ),
        # command_mode config
        other=dict(
            # Epsilon greedy with decay.
            eps=dict(
                # Decay type. Support ['exp', 'linear'].
                type='exp',
                start=0.95,
                end=0.1,
                decay=50000,
            ),
            replay_buffer=dict(replay_buffer_size=10000000, )
        ),
    ),
)
stocks_dqn_config = EasyDict(stocks_dqn_config)
main_config = stocks_dqn_config

stocks_dqn_create_config = dict(
    env=dict(
        type='methanol_futures-v0',
        import_names=['dizoo.gym_anytrading.envs.methanol_futures_env'],
    ),
    env_manager=dict(type='base'),
    policy=dict(
        type='dqn',
    ),
    evaluator=dict(
        type='trading_interaction',
        import_names=['dizoo.gym_anytrading.worker'],
    ),
    renderer=dict(
        type='plt',  # 使用plt渲染
        import_names=['dizoo.gym_anytrading.worker'],  # 导入渲染器路径
        evaluator=dict(
            force_render=True,  # 强制渲染
        ),
    ),
)
stocks_dqn_create_config = EasyDict(stocks_dqn_create_config)
create_config = stocks_dqn_create_config

if __name__ == "__main__":
    from ding.entry import serial_pipeline
    serial_pipeline([main_config, create_config], seed=0, max_env_step=int(1e7))
`
<img width="782" alt="Snipaste_2023-07-26_14-43-50" src="https://github.com/opendilab/DI-engine/assets/43986624/1cdd7266-2802-4286-a5ba-325856a87151">
<img width="778" alt="Snipaste_2023-07-26_14-44-27" src="https://github.com/opendilab/DI-engine/assets/43986624/769372ba-cf60-46cb-8c85-b568b711a362">
<img width="842" alt="Snipaste_2023-07-26_14-50-23" src="https://github.com/opendilab/DI-engine/assets/43986624/f550b54b-80ba-4a59-9a0b-ec4fbec18cba">
<img width="972" alt="Snipaste_2023-07-26_15-09-35" src="https://github.com/opendilab/DI-engine/assets/43986624/c82e73ee-e564-4f36-92a6-b915d7b13cf4">

**Comment by Cloud-Pku** [comment]
According to the logs, eps_len is indeed large enough to take up too much memory.

Evaluator prints out that your accumulated reward for each test is 0, I see you replaced 'base_env' with 'dizoo.gym_anytrading.envs.methanol_futures_env', is there any modification in it? It feels like something is wrong with '_calculate_reward' in methanol_futures_env.

Then, when the agent collects, the reward received for each action is also 0, resulting in the situation in the result diagram.

Maybe you can use tensorboard to show the training curve, which has a more detailed training status.

**Comment by 123xian123** [comment]
This is my "methanol_futures_env.py" file. In fact, I only modified the "selected_feature" part to fit my data. Below is a screenshot of my data. I noticed that there are negative values in the "buying" column, and I wonder if this could have an impact. The model has been trained for a long time, but the profit is still 1.
Regarding the use of TensorBoard, I will give it a try.
`from typing import Any
from copy import deepcopy
import numpy as np
from dizoo.gym_anytrading.envs.trading_env import TradingEnv, Actions, Positions, load_dataset
from ding.utils import ENV_REGISTRY



@ENV_REGISTRY.register('methanol_futures-v0')
class MethanolFuturesEnv(TradingEnv):

    def __init__(self, cfg):

        super().__init__(cfg)

        # ====== load Google stocks data =======
        raw_data = load_dataset(self._cfg.methanol_futures_data_filename, 'time')
        self.raw_prices = raw_data.loc[:, 'price'].to_numpy()
        EPS = 1e-10
        self.df = deepcopy(raw_data)
        if self.train_range == None or self.test_range == None:
            self.df = self.df.apply(lambda x: (x - x.mean()) / (x.std() + EPS), axis=0)
        else:
            boundary = int(len(self.df) * self.train_range)
            train_data = raw_data[:boundary].copy()
            boundary = int(len(raw_data) * (1 + self.test_range))
            test_data = raw_data[boundary:].copy()

            train_data = train_data.apply(lambda x: (x - x.mean()) / (x.std() + EPS), axis=0)
            test_data = test_data.apply(lambda x: (x - x.mean()) / (x.std() + EPS), axis=0)
            self.df.loc[train_data.index, train_data.columns] = train_data
            self.df.loc[test_data.index, test_data.columns] = test_data
        # ======================================

        # set cost
        self.trade_fee_bid_percent = 0.001  # unit
        self.trade_fee_ask_percent = 0.001  # unit

    # override
    def _process_data(self, start_idx: int = None) -> Any:
        '''
        Overview:
            used by env.reset(), process the raw data.
        Arguments:
            - start_idx (int): the start tick; if None, then randomly select.
        Returns:
            - prices: the close.
            - signal_features: feature map
            - feature_dim_len: the dimension length of selected feature
        '''

        # ====== build feature map ========
        all_feature_name = ['price', 'position', 'buying', 'vol', 'open', 'close', 'type']
        all_feature = {k: self.df.loc[:, k].to_numpy() for k in all_feature_name}
        # add feature "Diff"
        prices = self.df.loc[:, 'price'].to_numpy()
        diff = np.insert(np.diff(prices), 0, 0)
        all_feature_name.append('Diff')
        all_feature['Diff'] = diff
        # =================================

        # you can select features you want
        selected_feature_name = ['price', 'Diff', 'position', 'buying', 'vol', 'open', 'close', 'type']
        # selected_feature_name = ['price', 'Diff', 'position']
        selected_feature = np.column_stack([all_feature[k] for k in selected_feature_name])
        feature_dim_len = len(selected_feature_name)

        # validate index
        if start_idx is None:
            if self.train_range == None or self.test_range == None:
                self.start_idx = np.random.randint(self.window_size, len(self.df) - self._cfg.eps_length)
            elif self._env_id[-1] == 'e':
                boundary = int(len(self.df) * (1 + self.test_range))
                assert len(self.df) - self._cfg.eps_length > boundary + self.window_size,\
                 "parameter test_range is too large!"
                self.start_idx = np.random.randint(boundary + self.window_size, len(self.df) - self._cfg.eps_length)
            else:
                boundary = int(len(self.df) * self.train_range)
                assert boundary - self._cfg.eps_length > self.window_size,\
                 "parameter test_range is too small!"
                self.start_idx = np.random.randint(self.window_size, boundary - self._cfg.eps_length)
        else:
            self.start_idx = start_idx

        self._start_tick = self.start_idx
        self._end_tick = self._start_tick + self._cfg.eps_length - 1

        return prices, selected_feature, feature_dim_len

    # override
    def _calculate_reward(self, action: int) -> np.float32:
        step_reward = 0.
        current_price = (self.raw_prices[self._current_tick])
        last_trade_price = (self.raw_prices[self._last_trade_tick])
        ratio = current_price / last_trade_price
        cost = np.log((1 - self.trade_fee_ask_percent) * (1 - self.trade_fee_bid_percent))

        if action == Actions.BUY and self._position == Positions.SHORT:
            step_reward = np.log(2 - ratio) + cost

        if action == Actions.SELL and self._position == Positions.LONG:
            step_reward = np.log(ratio) + cost

        # if action == Actions.DOUBLE_SELL and self._position == Positions.LONG:
        #     step_reward = np.log(ratio) + cost
        #
        # if action == Actions.DOUBLE_BUY and self._position == Positions.SHORT:
        #     step_reward = np.log(2 - ratio) + cost

        step_reward = float(step_reward)

        return step_reward

    # override
    def max_possible_profit(self) -> float:
        current_tick = self._start_tick
        last_trade_tick = current_tick - 1
        profit = 1.

        while current_tick <= self._end_tick:

            if self.raw_prices[current_tick] < self.raw_prices[current_tick - 1]:
                while (current_tick <= self._end_tick
                       and self.raw_prices[current_tick] < self.raw_prices[current_tick - 1]):
                    current_tick += 1

                current_price = self.raw_prices[current_tick - 1]
                last_trade_price = self.raw_prices[last_trade_tick]
                tmp_profit = profit * (2 - (current_price / last_trade_price)) * (1 - self.trade_fee_ask_percent
                                                                                  ) * (1 - self.trade_fee_bid_percent)
                profit = max(profit, tmp_profit)
            else:
                while (current_tick <= self._end_tick
                       and self.raw_prices[current_tick] >= self.raw_prices[current_tick - 1]):
                    current_tick += 1

                current_price = self.raw_prices[current_tick - 1]
                last_trade_price = self.raw_prices[last_trade_tick]
                tmp_profit = profit * (current_price / last_trade_price) * (1 - self.trade_fee_ask_percent
                                                                            ) * (1 - self.trade_fee_bid_percent)
                profit = max(profit, tmp_profit)
            last_trade_tick = current_tick - 1

        return profit

    def __repr__(self) -> str:
        return "DI-engine Methanol Futures Trading Env"`

<img width="440" alt="Snipaste_2023-07-26_16-26-28" src="https://github.com/opendilab/DI-engine/assets/43986624/9995a879-fe15-42bd-859d-dbafefac9999">

**Comment by Cloud-Pku** [comment]
These features need to be normalized before using. I think the input deviation is so large that the network always outputs an extreme value. So, the agent always takes Short.
Beside normalization, maybe you can print the 'step_reward, action, self._position' in _calculate_reward function during running. Then we can check if the logic of environment is right.
